File: Java/ChatGPT4/NBodySimulationPar.java
import java.util.Random;

public class NBodySimulationPar {

    static final int N_BODIES = 10_000;
    static final int STEPS = 100;
    static final double G = 6.67430e-11;
    static final double DT = 1.0;

    static class Body {
        double x, y, z;
        double vx, vy, vz;
        double ax, ay, az;
        double mass;

        public Body(double x, double y, double z, double vx, double vy, double vz, double mass) {
            this.x = x; this.y = y; this.z = z;
            this.vx = vx; this.vy = vy; this.vz = vz;
            this.mass = mass;
        }
    }

    static Body[] bodies;

    public static void initializeSystem(int n) {
        bodies = new Body[n + 1];
        Random rand = new Random(42);

        // Central body
        double centralMass = 1e20;
        bodies[0] = new Body(0, 0, 0, 0, 0, 0, centralMass);

        double radius = 1e7;

        for (int i = 1; i <= n; i++) {
            double angle = 2 * Math.PI * i / n;
            double r = radius * (1.0 + 0.1 * rand.nextDouble());

            double x = r * Math.cos(angle);
            double y = r * Math.sin(angle);
            double z = 0;

            double v = Math.sqrt(G * centralMass / r);
            double vx = -v * Math.sin(angle);
            double vy = v * Math.cos(angle);
            double vz = 0;

            bodies[i] = new Body(x, y, z, vx, vy, vz, 1.0);
        }
    }

    public static void computeForces() {
        for (Body b : bodies) {
            b.ax = b.ay = b.az = 0;
        }
    
        int n = bodies.length;
    
        java.util.Arrays.stream(bodies).parallel().forEach(bi -> {
            for (Body bj : bodies) {
                if (bi == bj) continue;
    
                double dx = bj.x - bi.x;
                double dy = bj.y - bi.y;
                double dz = bj.z - bi.z;
    
                double distSqr = dx * dx + dy * dy + dz * dz + 1e-10;
                double dist = Math.sqrt(distSqr);
                double force = G * bj.mass / (distSqr * dist);
    
                bi.ax += dx * force;
                bi.ay += dy * force;
                bi.az += dz * force;
            }
        });
    }
    // This parallel version is not symmetric, so it's about twice as expensive as the original i < j version, but parallel and simpler to implement. Further optimization may require fine-grained task splitting.

    public static void kickStep() {
        for (Body b : bodies) {
            b.vx += b.ax * DT;
            b.vy += b.ay * DT;
            b.vz += b.az * DT;

            b.x += b.vx * DT;
            b.y += b.vy * DT;
            b.z += b.vz * DT;
        }
    }

    public static double computeEnergy() {
        double kinetic = java.util.Arrays.stream(bodies).parallel().mapToDouble(b -> {
            double v2 = b.vx * b.vx + b.vy * b.vy + b.vz * b.vz;
            return 0.5 * b.mass * v2;
        }).sum();
    
        int n = bodies.length;
        double potential = java.util.stream.IntStream.range(0, n).parallel().mapToDouble(i -> {
            Body bi = bodies[i];
            double sum = 0;
            for (int j = i + 1; j < n; j++) {
                Body bj = bodies[j];
                double dx = bj.x - bi.x;
                double dy = bj.y - bi.y;
                double dz = bj.z - bi.z;
                double dist = Math.sqrt(dx * dx + dy * dy + dz * dz + 1e-10);
                sum -= G * bi.mass * bj.mass / dist;
            }
            return sum;
        }).sum();
    
        return kinetic + potential;
    }
    
    public static void main(String[] args) {
        long start = System.currentTimeMillis();
        System.out.println("Initializing...");
        initializeSystem(N_BODIES);

        System.out.println("Computing initial energy...");
        double initialEnergy = computeEnergy();
        System.out.printf("Initial total energy: %.6e\n", initialEnergy);

        for (int step = 0; step < STEPS; step++) {
            computeForces();
            kickStep();

            if (step % 100 == 0) {
                System.out.println("Step " + step + " complete");
            }
        }

        System.out.println("Computing final energy...");
        double finalEnergy = computeEnergy();
        System.out.printf("Final total energy: %.6e\n", finalEnergy);
        System.out.printf("Energy difference: %.6e\n", Math.abs(finalEnergy - initialEnergy));

        long end = System.currentTimeMillis();
        System.out.printf("Simulation completed in %.2f seconds.\n", (end - start) / 1000.0);
    }
}


// Below are only the modified sections from the previous code (to keep it concise):
// Replace computeForces() with:


// Replace computeEnergy() with:


// You must run this with adequate heap space, like:
// java -Xmx16g NBodySimulation
// Java Parallel Streams scale well up to a few dozen cores. Beyond that, you'll want finer control (e.g., with ForkJoinPool or ExecutorService).
// This is still O(N²). For real performance at scale, consider a Barnes-Hut tree or Fast Multipole Method.

File: Julia/ChatGPT4/NBodySimulationPar.jl
using Random
using Printf
using Base.Threads

const G = 6.67430e-11
const DT = 1.0
const N_BODIES = 10_000
const STEPS = 100

mutable struct Body
    x::Float64
    y::Float64
    z::Float64
    vx::Float64
    vy::Float64
    vz::Float64
    ax::Float64
    ay::Float64
    az::Float64
    mass::Float64
end

function initialize_system(n)
    bodies = Vector{Body}(undef, n + 1)

    bodies[1] = Body(0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1e20)
    radius = 1e7
    rng = MersenneTwister(42)

    for i in 2:n+1
        angle = 2π * (i - 1) / n
        r = radius * (1.0 + 0.1 * rand(rng))
        x = r * cos(angle)
        y = r * sin(angle)
        z = 0.0
        v = sqrt(G * bodies[1].mass / r)
        vx = -v * sin(angle)
        vy = v * cos(angle)
        vz = 0.0
        bodies[i] = Body(x, y, z, vx, vy, vz, 0.0, 0.0, 0.0, 1.0)
    end

    return bodies
end

function compute_forces!(bodies)
    n = length(bodies)

    @threads for i in 1:n
        bi = bodies[i]
        ax = ay = az = 0.0

        for j in 1:n
            if i == j
                continue
            end
            bj = bodies[j]
            dx = bj.x - bi.x
            dy = bj.y - bi.y
            dz = bj.z - bi.z
            dist2 = dx^2 + dy^2 + dz^2 + 1e-10
            dist = sqrt(dist2)
            force = G * bj.mass / (dist2 * dist)
            ax += dx * force
            ay += dy * force
            az += dz * force
        end

        bi.ax = ax
        bi.ay = ay
        bi.az = az
    end
end

function update_bodies!(bodies)
    Threads.@threads for i in eachindex(bodies)
        b = bodies[i]
        b.vx += b.ax * DT
        b.vy += b.ay * DT
        b.vz += b.az * DT
        b.x += b.vx * DT
        b.y += b.vy * DT
        b.z += b.vz * DT
    end
end

function compute_energy(bodies)
    n = length(bodies)
    kinetic = zeros(Float64, n)
    potential = zeros(Float64, n)

    Threads.@threads for i in 1:n
        bi = bodies[i]
        kinetic[i] = 0.5 * bi.mass * (bi.vx^2 + bi.vy^2 + bi.vz^2)

        pot = 0.0
        for j in i+1:n
            bj = bodies[j]
            dx = bi.x - bj.x
            dy = bi.y - bj.y
            dz = bi.z - bj.z
            dist = sqrt(dx^2 + dy^2 + dz^2 + 1e-10)
            pot -= G * bi.mass * bj.mass / dist
        end
        potential[i] = pot
    end

    return sum(kinetic) + sum(potential)
end

function main()
    println("Initializing bodies...")
    bodies = initialize_system(N_BODIES)

    println("Computing initial energy...")
    energy0 = compute_energy(bodies)
    @printf("Initial energy: %.6e\n", energy0)

    for step in 1:STEPS
        compute_forces!(bodies)
        update_bodies!(bodies)
        if step % 100 == 0
            println("Step $step")
        end
    end

    println("Computing final energy...")
    energy1 = compute_energy(bodies)
    @printf("Final energy: %.6e\n", energy1)
    @printf("Energy difference: %.6e\n", abs(energy1 - energy0))
end

main()
